src/main/java/com/google/firebase/internal/ApacheHttp2Transport.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2024 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import com.google.api.client.http.HttpTransport;
+
+import java.io.IOException;
+import java.net.ProxySelector;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.hc.client5.http.async.HttpAsyncClient;
+import org.apache.hc.client5.http.async.methods.SimpleRequestBuilder;
+import org.apache.hc.client5.http.config.TlsConfig;
+import org.apache.hc.client5.http.impl.async.CloseableHttpAsyncClient;
+import org.apache.hc.client5.http.impl.async.HttpAsyncClientBuilder;
+import org.apache.hc.client5.http.impl.nio.PoolingAsyncClientConnectionManager;
+import org.apache.hc.client5.http.impl.routing.SystemDefaultRoutePlanner;
+import org.apache.hc.core5.http.config.Http1Config;
+import org.apache.hc.core5.http2.HttpVersionPolicy;
+import org.apache.hc.core5.http2.config.H2Config;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.hc.core5.util.TimeValue;
+
+public final class ApacheHttp2Transport extends HttpTransport {
+
+  private final CloseableHttpAsyncClient httpAsyncClient;
+  private final boolean isMtls;
+
+  public ApacheHttp2Transport() {
+    this(newDefaultHttpAsyncClient(), false);
+  }
+
+  public ApacheHttp2Transport(CloseableHttpAsyncClient httpAsyncClient) {
+    this(httpAsyncClient, false);
+  }
+
+  public ApacheHttp2Transport(CloseableHttpAsyncClient httpAsyncClient, boolean isMtls) {
+    this.httpAsyncClient = httpAsyncClient;
+    this.isMtls = isMtls;
+    
+    httpAsyncClient.start();
+  }
+
+  public static CloseableHttpAsyncClient newDefaultHttpAsyncClient() {
+    return defaultHttpAsyncClientBuilder().build();
+  }
+
+  public static HttpAsyncClientBuilder defaultHttpAsyncClientBuilder() {
+    PoolingAsyncClientConnectionManager connectionManager = 
+        new PoolingAsyncClientConnectionManager();
+    connectionManager.setMaxTotal(100);
+    connectionManager.setDefaultMaxPerRoute(100);
+    connectionManager.closeIdle(TimeValue.of(30, TimeUnit.SECONDS));
```

Comment:
Should we add a note on how we decided on these limits? 100 max connections and 30 seconds timeout?

Reply:
Good point, I changed the settings here to match the google api client's apache http transport and added a comment stating that. 

Also correctly limited the max concurrent streams to 100 and added a comment on the reasoning for this.

Also added more additional comments throughout this file.